tests/components/dlna_dms/conftest.py

```diff
@@ -129,3 +128,41 @@ def ssdp_scanner_mock() -> Iterable[Mock]:
         reg_callback = mock_scanner.return_value.async_register_callback
         reg_callback.return_value = Mock(return_value=None)
         yield mock_scanner.return_value
+
+
+@pytest.fixture
+async def device_source_mock(
+    hass: HomeAssistant,
+    config_entry_mock: MockConfigEntry,
+    ssdp_scanner_mock: Mock,
+    dms_device_mock: Mock,
+) -> AsyncIterable[DmsDeviceSource]:
+    """Fixture to set up a DmsDeviceSource in a connected state and cleanup at completion."""
+    config_entry_mock.add_to_hass(hass)
+    assert await hass.config_entries.async_setup(config_entry_mock.entry_id)
+    await hass.async_block_till_done()
+
+    domain_data = cast(DlnaDmsData, hass.data[DOMAIN])
```

Comment:
We shouldn't need to access hass.data in the tests. We don't want to interact with any integration details.

We can assert core state, library mocks or eg return values from the media_source integration top level api.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

Reply:
I think I get where you're coming from. We want to black-box test the integrations as much as possible.

However, there's a few parts in `dlna_dms` where the test is done better by accessing the internal data directly. For example, checking invariants related to `DmsDeviceSource.source_id` and `DlnaDmsData.sources` is a good thing to do during the name change test, but can't be done without accessing the domain data. Most integrations wouldn't need to worry about that because they have entities handled by entity platforms. [Media sources are not entities](https://github.com/home-assistant/architecture/discussions/675#discussioncomment-1744709), so some things are done manually in `dlna_dms`.

I've changed most tests where it can be done via the `media_source` interface, but left some using the domain data directly for reasons outlined above.

Reply:
I don't agree. How is the name change expressed in core state, library interface or media source interface or other interfaces that matter.

Reply:
I think that you should add motivations and explanations to the "Writing tests for integrations" section. Currently the only reason is "Following this pattern will make the tests more robust for changes in the integration", but it doesn't say how following the commandments will accomplish that. How is using `hass.data` going to break things? Why not unit-test the integration components, as well as black-box test the integration as a whole?

To rewrite `test_dlna_dms_data.py::test_update_source_id` so it doesn't touch `hass.data` I would have to:
1. Add a 2nd config entry to change the behaviour of `DmsMediaSource.async_browse_media` so that I could get the source ID(s) out via the `media_player.async_browse_media` call.
2. Call `media_player.async_browse_media` on the old and new source IDs to see that the dict keys have shifted
3. Catch the resulting `BrowseError`s and match on their strings to see if the error is because the source ID doesn't exist or because it does exist but the device mock is "disconnected".

This makes it more fragile, as now the test relies on the exact top-level browsing behaviour (1 above) and the exact error message strings (3). And after all that I still wouldn't know if there's any bad keys added to `DlnaDmsData.sources` by the name-change algorithm, because the test can't peek into the dict.

The only interfaces available are `media_source.async_browse_media` and `media_player.async_resolve_media`. There's no core state. The library doesn't know nor care what we call the sources.

Reply:
I looked a bit at the source id code. I think the main problem is that we use the config entry title as a source identifier instead of using a stable id. Why do we not use the config entry entry_id as eg the motion_eye integration does?

Reply:
For `dlna_dms`, the [`source_id`s](https://www.home-assistant.io/integrations/dlna_dms/#media-source-uris) are used like an `entity_id` would be in most other service calls: to distinguish which entity (DMS server) is being used. They are user-visible and can be used in `media_player.play_media` service calls. I made sure that media items could be specified in the service call without having to select the item via the HA frontend, and that all parts of the `media-source://dlna_dms/` URI could be hand-written without knowledge of UUIDs.

I had a look at the motionEye integration. Its media source capabilities seem to be mostly about providing a browsable view of captured events within the frontend, or taking events and piping stored file URLs directly to a media player.

Compare `"media-source://dlna_dms/my_server/videos/favourites/Epic Sax Guy 10 Hours.mp4` to `media-source://motioneye/74565ad414754616000674c87bdc876c#662aa1c77657dbc4af836abcdf80000a#movies#/2021-04-10/21-27-53.mp4` for someone trying to generate the URL in a `media_player.play_media` service call.

Reply:
We can't use user generated content (config entry title) in any identifier that is meant to be stable. Now renaming config entry has unintended side effects and causes frustration.

If you want a stable identifier, you could take the server name, slugify it and store it in config entry data and then use that as a stable identifier (which cannot be changed).

Reply:
Ignoring the fact that [`entity_id` can be changed](https://www.home-assistant.io/docs/configuration/customizing-devices/#changing-the-entity-id), I've solidified the `source_id`s in the config_entry and written code to add them to existing entries.

Reply:
I've also reduced the use of `hass.data` to checking for resource clean up, and for one test where it's too hard to get the timing right via SSDP and asyncio.